src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -298,19 +298,29 @@ abstract class BTypes {
    * referring to BTypes.
    */
   sealed trait BType {
-    final override def toString: String = this match {
-      case UNIT   => "V"
-      case BOOL   => "Z"
-      case CHAR   => "C"
-      case BYTE   => "B"
-      case SHORT  => "S"
-      case INT    => "I"
-      case FLOAT  => "F"
-      case LONG   => "J"
-      case DOUBLE => "D"
-      case ClassBType(internalName) => "L" + internalName + ";"
-      case ArrayBType(component)    => "[" + component
-      case MethodBType(args, res)   => "(" + args.mkString + ")" + res
+    final override def toString: String = {
```

Comment:
@lrytz split out several PRs for this https://github.com/scala/scala/pull/5831 https://github.com/scala/scala/pull/5834 https://github.com/scala/scala/pull/5833 https://github.com/scala/scala/pull/5832

haven't test them or got he performance figures - will run overnight and post up if successful